bson-record-codec/src/main/org/bson/codecs/record/annotations/BsonProperty.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.record.annotations;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * An annotation that configures a record component.
+ *
+ * @since 4.XXX
+ */
+@Documented
+@Target({ElementType.RECORD_COMPONENT})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface BsonProperty {
+    /**
+     * The field name of the record component.
+     *
+     * @return the field name to use for the record component
+     */
+    String value() default "";
```

Comment:
Empty string is actually allowed as a field name, so I don't think it's necessary or correct to have this check.  The bigger problem is if multiple components end up with the same name.  It would be trivial to check for that, but I don't think POJO codec does either.

Reply:
> Empty string is actually allowed as a field name

Unexpected, but cool) Nevertheless, do you think that specifying the default value for `BsonProperty.value` is actually useful? To me it seems that the API should push users into providing a field name, rather than tolerating no explicit name. The latter allows to specify multiple `@BsonProperty` annotations with no explicit value, i.e., with the same empty field name.

Reply:
I think that it was done this way on the existing BsonProperty annotation so that if you just wanted to specify `useDiscriminator` you could do so, and empty string for `value` was just ignored.  I can remove the default for now, but we might want it in the future if we add more properties to the annotation.

Reply:
> I can remove the default for now, but we might want it in the future if we add more properties to the annotation.

Adding the empty default in the future and start ignoring it (if we don't then it defeats the purpose in the scenario you described) will be a breaking change.

> the existing BsonProperty annotation ... empty string for value was just ignored

Since the POJO codec does not allow using an empty string (ignores the value and uses the name of the Java field), maybe `RecordCodec` should have the same behavior? I know I mentioned not allowing empty names earlier, and you said they are allows, but with the new information you provided, it seems like `RecordCodec` behaves differently from the POJO codec, and I am not sure it's a good thing.